[Hashicorp] Enable ADC for Vault GCP authentication

fpopic opened the pull request:
### Description

This PR consolidates the GCP authentication logic within the Hashicorp Vault secrets backend. Previously, the `gcp` authentication type strictly required a service account key file. This change enables the `gcp` type to automatically fallback to Application Default Credentials (ADC) when no key is provided, making it compatible with Google Cloud Composer, GKE, and GCE managed identities.

GCP auth method reference in Hashicorp Vault docs: https://developer.hashicorp.com/vault/docs/auth/gcp#google-cloud-auth-method

**Key Changes:**
- **Unified `gcp` Auth**: The `gcp` authentication type now handles both key-based and keyless (managed identity) authentication flows.
- **Robust Identity Detection**: Added logic to reliably determine the service account email from credentials, including a fallback refresh for Compute Engine credentials.
- **JWT Bug Fix**: Corrected the `sub` claim in the signed JWT payload. It now correctly passes the service account email string instead of the credentials object.
- **Documentation**: Updated `VaultHook` and `VaultBackend` docstrings to reflect the support for `gcp` ADC and the missing `jwt` auth type.
- **Tests**: Added `test_gcp_adc` and updated existing GCP tests to verify the unified flow and identity extraction.

---

##### Was generative AI tooling used to co-author this PR?

- [X] Yes (Gemini CLI 3.0)
- [X] Yes (Codex CLI gpt-5.5)

Generated-by: Gemini CLI following [the guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#gen-ai-assisted-contributions)

---

* related: #51991

fpopic commented:
Will fix unit tests.

fpopic committed:
[Hashicorp] Enable ADC for Vault GCP authentication

This commit unifies and improves the GCP authentication logic within the Hashicorp Vault secrets backend.
The 'gcp' authentication type now supports both service account key-based authentication
and Application Default Credentials (ADC) automatically.

Changes:
- Unified GCP Authentication: The 'gcp' auth type now handles both key-based and keyless (ADC) flows.
- Automatically determine service account email from credentials, with a fallback
  to the GCE/GKE metadata server.
- Fix a bug where the 'sub' claim in the signed JWT payload was incorrectly
  set to the credentials object instead of the service account email.
- Update VaultHook and VaultBackend docstrings and initialization logic.
- Add unit tests for ADC-based GCP authentication and update existing tests
  (test_gcp renamed to test_gcp_key) to verify the corrected identity extraction.

fpopic committed:
[Hashicorp] Fix linting and formatting in Vault GCP auth

Address static check failures (prek) by fixing whitespace issues,
correcting docstring typos, and applying standard formatting
to the Vault GCP authentication changes.

- Fix W293 (blank line contains whitespace) in vault_client.py and tests.
- Correct typos and inconsistent quoting in _VaultClient docstrings.
- Apply ruff formatting for project consistency.

fpopic committed:
[Hashicorp] Update Vault GCP hook tests for ADC and service account email discovery

Refactor Vault GCP authentication tests in TestVaultHook to:
- Properly mock credentials object with 'service_account_email' attribute to match
  updated _VaultClient discovery logic.
- Remove redundant 'builtins.open' mocks in tests that no longer directly
  open key files, relying instead on the mocked credential provider.
- Fix VaultError: 'Could not determine service account email from credentials'
  in unit tests.

fpopic commented:
> @fpopic , Would a user ever not want this to happen? Do you think its worth providing a way to not automatically have the fallback to

@rawwar I checked also [aws_iam](https://github.com/fpopic/airflow/blob/9c8114f1f94463f86a5aae42d54ac04910551572/providers/hashicorp/src/airflow/providers/hashicorp/_internal_client/vault_client.py#L401-L427) auth type which follows the similar approach (e.g. checks for static credentials otherwise falls back assuming short-lived credentials).

One counter-example gemini cli brought:
> 
>   A common counter-example is environment isolation for security compliance.
> 
>   In some highly regulated environments (e.g., PCI-DSS or HIPAA), a single Airflow cluster might be used to run DAGs for multiple different business units. If a developer accidentally forgets to provide a service account key in their Vault connection, the system might silently fall back to the GKE/GCE Node's default identity.
>
>   If that node identity has broad permissions (which is unfortunately a common misconfiguration), the DAG could gain unauthorized access to secrets in Vault that it shouldn't have. In such cases, the user would prefer a "Fail Fast" behavior: the task should crash immediately if the explicitly requested credentials are missing, rather than defaulting to a "powerful" environment identity.

Since this pattern is already established for cloud, the concern about "silent fallback"
or GCP is effectively a concern about the provider's general architecture, which this PR correctly
brings into parity for Google Cloud.

WDYT?

fpopic commented on providers/hashicorp/src/airflow/providers/hashicorp/_internal_client/vault_client.py:
```diff
@@ -352,25 +348,38 @@ def _auth_gcp(self, _client: hvac.Client) -> None:
         now = int(time.time())
         expires = now + 900  # 15 mins in seconds, can't be longer.
-        payload = {"iat": now, "exp": expires, "sub": credentials, "aud": f"vault/{self.role_id}"}
+        payload = {"iat": now, "exp": expires, "sub": service_account_email, "aud": f"vault/{self.role_id}"}
```
> Hmmmm, what is the impact of this change? Was this just never working?

https://github.com/apache/airflow/issues/17500#:~:text=It%27s%20a%20bold%20statement%20as%20it%20means%20nobody%20has%20really%20used%20VaultBackend%20with%20these%20auth%20types.%20I%20may%20be%20wrong%2C%20but%20I%20would%20be%20surprised :

> It's a bold statement as it means nobody has really used VaultBackend with these auth types. I may be wrong, but I would be surprised.

Can totally be. 😓 

The previous implementation was passing the entire `credentials` object to `json.dumps()`, which resulted in a `TypeError` during serialization. This means the `gcp` auth type was likely non-functional for any signed JWT flow until now. I've corrected it to pass the service account email string, which aligns with the GCP `signJwt` API requirements.

potiuk commented:
@fpopic A few things need addressing before review — see our [Pull Request quality criteria](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-quality-criteria).

- :warning: **Unresolved review comments**: This PR has 2 unresolved review threads from maintainers: **@ashb** (MEMBER): 2 unresolved threads. Please review and resolve all inline review comments before requesting another review. You can resolve a conversation by clicking 'Resolve conversation' on each thread after addressing the feedback. See [pull request guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst).

> **Note:** Your branch is **300 commits behind `main`**. Please rebase and push again to get up-to-date CI results.

No rush.

---

_Note: This comment was drafted by an AI-assisted triage tool and may contain mistakes. Once you have addressed the points above, an Apache Airflow maintainer — a real person — will take the next look at your PR. We use this [two-stage triage process](https://github.com/apache/airflow/blob/main/contributing-docs/25_maintainer_pr_triage.md#why-the-first-pass-is-automated) so that our maintainers' limited time is spent where it matters most: the conversation with you._

fpopic committed:
[Hashicorp] Refine Vault GCP auth discovery and error handling

Address reviewer concerns by improving the robustness of service account
email detection and type-checking. Ensure the JWT 'sub' claim correctly
uses the service account email string instead of the credentials object.

- Use 'is None' check for identity discovery to avoid silent fallbacks.
- Include explicit type information in VaultError for easier debugging.
- Fix broken JWT payload in GCP authentication.
- Update unit tests to handle stricter type checks and improved mocks.

fpopic commented on providers/hashicorp/src/airflow/providers/hashicorp/_internal_client/vault_client.py:
```diff
@@ -352,25 +348,38 @@ def _auth_gcp(self, _client: hvac.Client) -> None:
+        if not service_account_email or not isinstance(service_account_email, str):
+            service_account_email = getattr(credentials, "client_email", None)
+
+        if not service_account_email or not isinstance(service_account_email, str):
```
I've updated the logic to use `if service_account_email is None:` for the fallback check. This ensures that if a value is explicitly provided but is of the wrong type, we don't silently fall back to instance discovery, which addresses your concern about unexpected behavior. I've also improved the error message to include the actual type received to make debugging easier.

fpopic commented:
I've rebased the branch on `main` and addressed all technical concerns raised by @ashb. The unit tests have been updated and expanded to cover the new type-checking logic and mock behavior. Ready for another review.

fpopic commented:
Re-request review @copilot due to bot unsuccessful review try.

fpopic committed:
Merge branch 'main' into fix_vault_gcp_auth_requirements

fpopic committed:
Merge branch 'main' into fix_vault_gcp_auth_requirements

fpopic committed:
Merge branch 'main' into fix_vault_gcp_auth_requirements

fpopic committed:
Merge branch 'main' into fix_vault_gcp_auth_requirements

fpopic committed:
[Hashicorp] Address PR feedback for Vault GCP authentication

Address feedback from reviewers and automated reviews:
- Update 'role_id' docstring in VaultHook, VaultBackend, and _VaultClient to include 'gcp' auth type.
- Synchronize JWT 'exp' claim to 900 seconds (15 minutes) in both implementation and tests to match Vault/GCP constraints.
- Refine exception handling during Compute Engine credential refresh to catch specific RefreshError instead of broad Exception.
- Improve unit tests by mocking 'time.time' for stable 'iat' and 'exp' assertions and removing hacky 'json.dumps' mocks.

fpopic commented:
I've addressed the latest feedback from Copilot (AI) and reviewers:
- Updated 'role_id' docstring in VaultHook, VaultBackend, and _VaultClient to include 'gcp' auth type.
- Synchronized JWT 'exp' claim to 900 seconds (15 minutes) in both implementation and tests.
- Refined exception handling during Compute Engine credential refresh to catch specific RefreshError.
- Improved unit tests by mocking 'time.time' for more stable assertions.

fpopic committed:
Merge branch 'main' into fix_vault_gcp_auth_requirements

fpopic committed:
Remove unused `import time` from test_vault_client.py

The refactor that replaced `int(time.time())` calls with
`@mock.patch("time.time")` decorators left the `import time` statement
dead — ruff flags it as F401.

fpopic committed:
Fix vault GCP ADC test formatting

fpopic committed:
Merge branch 'main' into fix_vault_gcp_auth_requirements
